### PR TITLE
Improving prebuild related auth errors

### PR DESCRIPTION
**Pull request by selfcontained**
## Description

This aims to improve the error messages we create and provide users when they attempt to enable prebuilds. It handles scenarios where we can't install the required webhooks for a myriad of reasons, and provides users actionable ways to resolve the errors.

Example of unauthorized repo error handling in the UI
| Needs to connect | Missing scopes | We're not sure |
|--------|--------|--------|
| image | image | image | 

I've also added error messaging for the case where it's not an unauthorized repo error, but still an error. This includes the error message we return.
image

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖[[deprecated]](https://githubnext.com/copilot-for-prs-sunset) Generated by Copilot at 26b2fec</samp>

This pull request improves the error handling and reporting for unauthorized access to repositories and prebuild webhooks across different SCM providers. It adds new fields and methods to the `RepositoryUnauthorizedError` message and class, and uses them to provide more information and feedback to the client. It also updates the error handling logic and types in various files and classes that deal with tokens, scopes, and clone URLs.

</details>

## Related Issue(s)

Fixes EXP-760

## How to test

* After signing up, import a repo
* Try and enable prebuilds. It should fail saying you need to reconnect with the provider. Click reconnect.
* It should complete the oauth popup flow and attempt to enable prebuilds automatically and succeed.
* Try going to your git provider user settings and removing necessary scopes (`repo` for github).
* Disable prebuilds on your repo and try to enable.
* Repeat the process above and it should result in enabled prebuilds.
* Bonus points for connecting Gitlab and doing the same flows. You'll need to import the repo, and then go remove the `api` scope from Gitlab and try to enable prebuilds so it can fail.

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - brad-prebub4353f6352</li>
	<li><b>🔗 URL</b> - <a href="https://brad-prebub4353f6352.preview.gitpod-dev.com/workspaces" target="_blank">brad-prebub4353f6352.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - brad-prebuild-auth-errors-gha.21813</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-brad-prebub4353f6352%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft preemptible
      Saves cost. Untick this only if you're really sure you need a non-preemtible machine.
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

**Comment by AlexTugarev** [comment]
That worked quite well. 🎉 





Screenshot 2023-12-13 at 08 09 44

Not sure about the message, should it refer to webhooks? 🤷🏻‍♂️ 
`Looks like we need to authorize with github.com to enabled prebuilds.`
vs.
`Looks like we need to authorize with github.com to create a webhook.`

**Comment by selfcontained** [comment]
> That worked quite well. 🎉
> 
> Screenshot 2023-12-13 at 08 09 44
> Not sure about the message, should it refer to webhooks? 🤷🏻‍♂️ `Looks like we need to authorize with github.com to enabled prebuilds.` vs. `Looks like we need to authorize with github.com to create a webhook.`

Yah this was prior to syncing with Cory on the UX there. It's now updated with messaging that isn't as specific as you suggest, but we felt will still guide users to fixing things. Suggestions welcome though.

File: components/dashboard/src/App.tsx [hunk 1]
```diff
@@ -26,20 +27,24 @@ const App: FC = () => {
...
     const search = useQueryParams();
+    const { isDark, setIsDark } = useTheme();
 
     useEffect(() => {
         const onKeyDown = (event: KeyboardEvent) => {
             if ((event.metaKey || event.ctrlKey) && event.key === "o") {
                 event.preventDefault();
                 history.push("/new");
                 return;
+            } else if (event.metaKey && event.ctrlKey && event.shiftKey && event.key === "M") {
```
Comment (selfcontained) [inline components/dashboard/src/App.tsx:38]: Added this back to just ease checking dark/light mode during development. Hopefully this combo of keys doesn't conflict with anything.

File: components/dashboard/src/index.css [hunk 2]
```diff
@@ -16,6 +16,9 @@
         /* Setup RGB color channel variables */
         --black: 22 22 22; /* #161616 */
         --white: 255 255 255; /* #FFFFFF */
+        /* TODO: determine if these are the reds we want - need to have correct constrast */
```
Comment (selfcontained) [inline components/dashboard/src/index.css:19]: Needed some red colors so set this up for now. The red in the podkit designs didn't have the right contrast so went with these for now (from tailwind reds).

File: components/dashboard/src/projects/ProjectSettings.tsx [hunk 3]
```diff
@@ -100,10 +129,36 @@ export default function ProjectSettingsView() {
...
+
+                if (newSettings.prebuilds?.enable && error?.code === ErrorCodes.NOT_AUTHENTICATED) {
+                    const { host, /*providerIsConnected, providerType,*/ repoName, requiredScopes } =
+                        error?.data as PartialMessage<RepositoryUnauthorizedError>;
+
+                    const toastId = `toast--host-authorized--${host}`;
+                    toast(
+                        <>
+                            <p>There was a problem enabling prebuilds on "{repoName}"</p>
+                            <p>Grant these permissions {JSON.stringify(requiredScopes)} and try again! 🙏🏻</p>
```
Comment (filiptronicek) [inline components/dashboard/src/projects/ProjectSettings.tsx:141]:
```suggestion
                            <p>Grant the {JSON.stringify(requiredScopes)} permissions and try again! 🙏🏻</p>
```

**Review by filiptronicek — APPROVED** [review]
Trying it out with GitHub and GitLab, the flow feels so much smoother now. Both with repos I should have access to and those I don't, I can prebuild away as expected.

Left some minor comments

**Review by AlexTugarev — APPROVED** [review]
🎉 Awesome!

**Pull request merged**